internal/controller/compute/user/user.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	userID := cr.Status.AtProvider.UserID
+
+	observed, resp, err := eu.service.UpdateUser(ctx, userID, cr.Spec.ForProvider)
+	if err != nil {
+		werr := errors.Wrap(err, errUserUpdate)
+		return managed.ExternalUpdate{}, compute.AddAPIResponseInfo(resp, werr)
+	}
+	conn := connectionDetails(cr, observed)
+
+	err = updateGroups(ctx, eu, userID, cr.Status.AtProvider.GroupIDs, cr.Spec.ForProvider.GroupIDs)
+	if err != nil {
+		return managed.ExternalUpdate{ConnectionDetails: conn}, errors.Wrap(err, "failed to update groups")
+	}
+
+	return managed.ExternalUpdate{ConnectionDetails: conn}, nil
+}
+
+func (eu *externalUser) Delete(ctx context.Context, mg resource.Managed) error {
+	user, ok := mg.(*v1alpha1.User)
+	if !ok {
+		return errors.Wrap(errors.New(errNotUser), "delete error")
+	}
+	user.SetConditions(xpv1.Deleting())
+
+	userID := user.Status.AtProvider.UserID
+
+	for _, groupID := range user.Status.AtProvider.GroupIDs {
+		if err := eu.service.DeleteUserFromGroup(ctx, groupID, userID); err != nil {
+			return errors.Wrap(err, "failed to remove user from a group")
+		}
+	}
+
+	resp, err := eu.service.DeleteUser(ctx, userID)
+	return compute.ErrorUnlessNotFound(resp, errors.Wrap(err, errUserDelete))
+}
+
+// updateGroups adds or remove groups for the userID.
+func updateGroups(ctx context.Context, eu *externalUser, userID string, atProviderGroups []string, forProviderGroups []string) error {
+	for _, groupID := range atProviderGroups {
+		if groupID != "" && !slices.Contains(forProviderGroups, groupID) {
+			if derr := eu.service.DeleteUserFromGroup(ctx, groupID, userID); derr != nil {
+				return errors.Wrap(derr, "failed to remove user from a group")
+			}
+		}
+	}
+
+	for _, groupID := range forProviderGroups {
+		_, _, gerr := eu.service.AddUserToGroup(ctx, groupID, userID)
+		// skip if user is already member of this group.
+		if gerr != nil && strings.Contains(gerr.Error(), "is already member of") {
+			continue
+		}
+		if gerr != nil {
+			return errors.Wrap(gerr, errAddUserToGroup)
+		}
+	}
+	return nil
+}
+
+func createMap(s []string) map[string]bool {
+	m := make(map[string]bool)
+	for i := range s {
+		m[s[i]] = false
+	}
+	return m
+}
+func compareStringSlices(s1, s2 []string) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	m1 := createMap(s1)
+	m2 := createMap(s2)
+	for s := range m1 {
+		_, exist := m2[s]
+		if !exist {
+			return false
+		}
+	}
+	return true
+}
+
+// isUserUpToDate returns true if the User is up-to-date or false otherwise.
+func isUserUpToDate(params v1alpha1.UserParameters, observed ionosdk.User, observedGroups []string) bool { //nolint:gocyclo
+	if !observed.HasProperties() {
+		return false
+	}
+
+	// After creation the password is stored as a connection detail secret
+	// and removed from the cr. If the cr has a password it means
+	// the client wants to update it.
+	if params.Password != "" {
+		return false
+	}
+
+	if !compareStringSlices(params.GroupIDs, observedGroups) {
```

Comment:
While this is definitely better than the reflect.DeepEqual way, wouldn't slices.Equal that was added in go 1.21(i think?) solve this?

Reply:
To run slices.Equal, we would need to run `slices.Sort` on both slices first. But I agree it would be a bit cleaner. This solution would be O(n log(n)) which is a bit more than Mohammed's O(n) comparison.
But we're working with fairly small slices were Mohammed's algo performs worst because of the `times 3` constant factor for allocating the sets and then iterating over them.